### DISCUSSION TITLE
Plugin dev: editing html files in `build` folder with `postBuild` not working

### DISCUSSION THREAD

**Discussion by 0xCaso** [facebook/docusaurus#8819] [Q&A] [answered]
I want to edit the files produced from the build, for example removing the svgs with a certain class (note: html files are built from markdown files, this is a documentation website).

I'm doing it with a custom plugin, which uses the postBuild lifecycle. The thing is that the things are actually saved in the html file (when I run npm run build) but when I run npm run serve I cannot see (rendered) the changes.

The weirdest thing is that I actually can see the changes from the browser, with the inspector, in the "Sources" section, but in the "Elements" section there is still the image.

Do you know what is wrong / what I'm missing?

Snippet:

```typescript
export default function graphvizSvgFixer(
  context: LoadContext,
): Plugin {
  return {
    name: 'graphviz-svg-fixer',
    async postBuild(props) {
        for (let htmlFile of htmlFilesToFix) {
          let html = fs.readFileSync(`${props.outDir}/${htmlFile.route}/index.html`);
          const $ = cheerio.load(html);
          const prev = $.html();
          $('svg.graphviz').remove();
          let next = $.html();
          console.log(prev == next);
          fs.writeFileSync(`${props.outDir}/${htmlFile.route}/index.html`, $.html());
        }
    },
  }
};
```
Below, you can find the svg both rendered both in the "Elements" section:

<img width="1015" alt="Screenshot 2023-03-24 at 10 32 56" src="https://user-images.githubusercontent.com/22419450/227533606-62521686-6740-4da6-bfe3-9dab5d3e2652.png">

Here you can see that if you look for "graphviz" in the "Sources" section nothing gets found:

<img width="1176" alt="Screenshot 2023-03-24 at 10 33 22" src="https://user-images.githubusercontent.com/22419450/227533613-74feb248-5167-4d17-802d-a8ef9daaab08.png">

Here, if I look for "Image: Block", always in "Sources", I find the place where the rendered svg should be, but it doesn't appear; it should be between "Image: Block" and "Definition ...":

<img width="1176" alt="Screenshot 2023-03-24 at 10 33 47" src="https://user-images.githubusercontent.com/22419450/227533620-087010da-16c4-4d57-882a-db85b6baa985.png">

**Comment by slorber** [comment]
That's impossible to help with just screenshots here. If `./build` is good, then there's no reason for `docusaurus serve` to not serve your static files. Double-check everything and you'll likely find out the problem

**Reply by 0xCaso** [reply]
> That's impossible to help with just screenshots here. If `./build` is good, then there's no reason for `docusaurus serve` to not serve your static files. Double-check everything and you'll likely find out the problem

Thanks. Any ideas about what to double-check? I've been checking the last two days honestly and can't figure out what's wrong. I also tried to see if it was a browser or cache issue, but nothing.

**Reply by slorber** [reply]
Check your `./build` folder. Remove everything in it except the html file and run `docusaurus serve` on it. Cleanup the html file to isolate the specific part and try again. If you end up with a minimalist html file and it doesn't serve the file content, then send me the minimalist html file and show me a proof (video recording) that the file content is not served on your computer.

**Reply by 0xCaso** [reply]
That's weird, I removed everything from `./build` and left only the html file. If I refresh it's without css, and the svgs are actually disappeared so it's rendering the correct thing I guess.

<img width="278" alt="Screenshot 2023-03-24 at 18 21 20" src="https://user-images.githubusercontent.com/22419450/227596516-16274a5c-0895-4433-8095-2ba25c6e741f.png">

If i leave something in the `.build` folder, the svg still render.
This is what I'm leaving:

<img width="680" alt="Screenshot 2023-03-24 at 18 24 42" src="https://user-images.githubusercontent.com/22419450/227597359-3cfd0cd9-3eb0-466b-a29a-d19d3d6a9e49.png">

The html page is inside the folder `chap-state`, which contains just that file, and the file hasn't the svg to render inside.

---

Update: if I remove the `assets` folder, the svg disappears. That folder contains only css and js stuff, the `svg` files are inside the `img` folder.

**Reply by 0xCaso** [reply]
Btw, I've opened the issue https://github.com/facebook/docusaurus/issues/8832 to understand if it's a docusaurus problem or mine.

**Reply by slorber** [reply]
We don't need another issue, this discussion is appropriate to figure out if it's a bug or not. 

If it's a bug, then you should be able to create a minimal repro and explain the problem properly. So far your explanations are quite messy and unstructured, and you assume that I understand your site which is not the case at all. 

Please do the upfront work to help me help you, otherwise, I won't.

**Comment by Josh-Cena** [comment]
Hi @0xCaso could you try running the webpage with JS disabled? Remember that Docusaurus hydrates the HTML with React, and the React code owns its own copy of what the DOM structure is like. If the plain HTML version works as intended, that means it's because React reverts your HTML changes after hydration.

One way to fix this is to move this logic to a Webpack plugin by intercepting SVG loading and modifying its source upfront.

**Reply by 0xCaso** [reply]
Hi @Josh-Cena, thanks a lot for the suggestion!

Disabling Javascript for the browser, I still see the "old" HTML file version, so without the update of my plugin.
In any case (with or without js enabled), when I load the website, I see the edited version for 1 millisecond, and that gets replaced by the old version.

I'll make a video soon if the situation is not clear, to explain it in the best way

**Reply by slorber** [reply]
@0xCaso Josh is probably right. Your imported SVGs are converted to React components that are rendered as HTML.

Editing the SVG output in the HTML is only one part of the story. 

The `svg.graphviz` elements still exists in the emitted React component if you look into the `./build/assets/**/*.js` files

You have to remove these elements not with a postBuild step modifying html files, but within the source React/svg code directly

It is not a bug in Docusaurus

> In any case (with or without js enabled), when I load the website, I see the edited version for 1 millisecond, and that gets replaced by the old version.

Try again, this seems really suspicious that there's this "flash" with JS disabled

**Reply by 0xCaso** [reply]
Thanks @slorber for the support!

What do you mean with this?
> You have to remove these elements not with a postBuild step modifying html files, but within the source React/svg code directly

Also I made a quick video to show the website behaviour: https://www.loom.com/share/a1a19c38f4354e7189296a7a06a50be2

As I said in the video, tell me if there are things that I should explain to make it more clear. Thanks again to both!

**Reply by slorber** [reply]
What you want is this:

<img width="538" alt="CleanShot 2023-04-07 at 10 17 04@2x" src="https://user-images.githubusercontent.com/749374/230571149-c7d3516f-d8fe-4e7b-b79d-20a8503f95d0.png">


But what React/JS renders is this:

<img width="534" alt="CleanShot 2023-04-07 at 10 17 33@2x" src="https://user-images.githubusercontent.com/749374/230571234-0d02d525-3dd4-477a-9c99-c3b1d95a7620.png">



You can't create plugins that just modify the html after a build like that. The HTML files will contain and show `-sec-num-`, but then the JS/React code loaded will "fix it" to the value it has in the source markdown.

If you want to modify anything that is rendered by React, it can't be done in a postBuild step: you have to find a way to make React render the final value in the first place. I can't help you doing so without you creating a small repro first.

Maybe this can help for the heading case

**Reply by 0xCaso** [reply]
That's clear @slorber, thanks. What do you mean with small repro? Should I create a new repo with just one example page and the plugins (so removing all the stuff that is useless for this case)?

(Also, both the SVG and headings problems are the same, as i always edit the HTML pages in postBuild, so the solution will solve both the problems)

**Reply by slorber** [reply]
> What do you mean with small repro? Should I create a new repo with just one example page and the plugins (so removing all the stuff that is useless for this case)?

You can use https://docusaurus.new and create the most minimal example of the problem, which I can easily run myself, and edit to give you an alternative solution with the problem fixed.

I don't even understand what you are attempting to do with these 2 plugins currently so it's hard to guide you in the right direction. I just know that your current approach is wrong

**Reply by 0xCaso** [reply]
The thing is that codesandbox.io hides the `build` folder, so there is no way to see the changes made by the plugins (you cannot delete files from the folder if it's hidden).
Would that still be useful for you? In case [here](https://codesandbox.io/p/github/0xCaso/polkadot-spec/draft/great-bush?file=README.md&workspace=%257B%2522activeFilepath%2522%253A%2522README.md%2522%252C%2522openFiles%2522%253A%255B%2522README.md%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522spaces%2522%253A%257B%2522clg6o4qex000x3b6kezl43twe%2522%253A%257B%2522key%2522%253A%2522clg6o4qex000x3b6kezl43twe%2522%252C%2522name%2522%253A%2522Default%2522%252C%2522devtools%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A3000%252C%2522key%2522%253A%2522clg6oxxyv01c93b6k970v6b9x%2522%252C%2522isMinimized%2522%253Afalse%252C%2522path%2522%253A%2522%252Fchap-state%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522key%2522%253A%2522clg6o56qt00353b6kdpu7ccmc%2522%252C%2522isMinimized%2522%253Atrue%257D%255D%257D%257D%252C%2522currentSpace%2522%253A%2522clg6o4qex000x3b6kezl43twe%2522%252C%2522spacesOrder%2522%253A%255B%2522clg6o4qex000x3b6kezl43twe%2522%255D%252C%2522hideCodeEditor%2522%253Afalse%257D) is the link to the sandbox, also if as for now it's basically the same of the actual repo. Otherwise I can create another GitHub repo with the minimal example, so you can test it locally.

**Reply by slorber** [reply]
>  Otherwise I can create another GitHub repo with the minimal example, so you can test it locally.

That's what I ask from the very beginning. If you throw me your big repo, you put the burden on me to investigate.

 If you want help, do the upfront work of removing everything that is not needed to the bare minimum, and describe your goal very concisely. The repro should show the failure you encounter, and nothing else.

**Reply by 0xCaso** [reply]
I'm sorry if I provide it through a GitHub repo, but as I said it's impossible through codesandbox.
This is the minimal example repo: https://github.com/0xCaso/polkadot-spec-min-example
I tried to explain what I want to achieve and the problem in the README.

Also, while I was writing the README, I thought about a possible solution (maybe it will be more understandable after reading the README):
- Put the existent markdown files (with the placeholders) in another dir (for example `/src/docs`)
- Make the plugin generate new markdown files to put inside `/docs` before the actual build
- Docusaurus during the build will generate HTML files from the new markdown files, which already contain the right content, so we have no issues with React stuff

This can solve the issue with the `numeration-system` plugin, but still I would have problems with the `graphviz-svg-fixer`, as for this one I'm forced to edit the generated HTML files (there is probably a way to avoid this, but it would be way more complex).

So the ideal thing would be to find a solution to both the problems, understanding how I can I say to React how to display the content of the edited HTML files, if this is the right way.

**Reply by 0xCaso** [reply]
FYI: I resolved the issue transforming the plugins into scripts, which I run before building the docusaurus website.

Btw, it would be still nice to know if it's possible to edit HTML after the build, in a way that React injections won't be a trouble.

**Reply by slorber** [reply, marked as answer]
> FYI: I resolved the issue transforming the plugins into scripts, which I run before building the docusaurus website.

Great, because your repro is still too complex for me to understand without spending hours on it. Thanks for your efforts but that's still not what I call a minimal repro of the problem

> Btw, it would be still nice to know if it's possible to edit HTML after the build, in a way that React injections won't be a trouble.

No it's not possible nor recommended with any React framework that has SSG/SSR (ie all of them except CRA which does not output anything in html pages, so... all of them). In general you should never do what you tried to do in any React context. 

Don't modify html content, but instead modify the React components that generate that html output.

**Reply by 0xCaso** [reply]
Thanks a lot for your time! In some way, the discussion was useful and I managed to make it, and hopefully this will save time to other people
